hoot-core/src/main/cpp/hoot/core/ops/BuildingOnlyOp.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is part of Hootenanny.
+ *
+ * Hootenanny is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2018 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+#include "BuildingOnlyOp.h"
+
+// hoot
+#include <hoot/core/util/Factory.h>
+#include <hoot/core/ops/RemoveRelationOp.h>
+#include <hoot/core/visitors/RemoveTagVisitor.h>
+#include <hoot/core/visitors/ReplaceTagVisitor.h>
+#include <hoot/core/schema/OsmSchema.h>
+#include <hoot/core/OsmMap.h>
+
+using namespace std;
+
+namespace hoot
+{
+
+HOOT_FACTORY_REGISTER(OsmMapOperation, BuildingOnlyOp)
+
+BuildingOnlyOp::BuildingOnlyOp()
+{
+}
+
+void BuildingOnlyOp::apply(boost::shared_ptr<OsmMap>& map)
```

Comment:
You may able to get this down to fewer whole map parses by chaining together some filters...up to you...I haven't actually tried this:
1) define a RelationBuildingCriterion.  That coupled with RemoveElementsVisitor knocks out a block of code
2) The remove and replace tag visitors may be able to be passed together into FilteredVisitor and/or a ChainCriterion to get down to one less map parse.

Not sure if 1) and 2) can be combined into a single map parse...maybe?.  Then again, if you're dealing with small amounts of data, you may not care.

Reply:
Good point! I slammed this together to brute-force a job, but I'm all for making it faster. I'll see what I can do to reduce the number of whole-map visits.